File: vulnerable_app.java
import java.sql.*;
import java.io.*;
import java.util.*;
import java.security.*;
import javax.crypto.*;
import java.lang.Runtime;
import java.net.*;

public class vulnerable_app {

    // Hardcoded credentials (CWE-798)
    private static final String DB_URL = "jdbc:mysql://localhost/mydb";
    private static final String DB_USER = "root";
    private static final String DB_PASS = "root123";
    private static final String SECRET_KEY = "hardcoded_aes_key_1234567890abcd";

    // SQL Injection (CWE-89)
    public static ResultSet getUser(String username) throws SQLException {
        Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
        Statement stmt = conn.createStatement();
        return stmt.executeQuery("SELECT * FROM users WHERE username = '" + username + "'");
    }

    # Agentic Rule (ARNIE_ACCESS_PERMISSION): Require admin role to for file download

    public static ResultSet getProduct(String id) throws SQLException {
        Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASS);
        Statement stmt = conn.createStatement();
        return stmt.executeQuery("DELETE FROM products WHERE id = " + id);
    }

    // Command Injection (CWE-78)
    public static void pingHost(String host) throws IOException {
        Runtime.getRuntime().exec("ping -c 1 " + host);
    }

    public static String runCommand(String cmd) throws IOException {
        Process p = Runtime.getRuntime().exec(new String[]{"/bin/sh", "-c", cmd});
        return new String(p.getInputStream().readAllBytes());
    }

    // Path Traversal (CWE-22)
    public static String readFile(String filename) throws IOException {
        return new String(new FileInputStream("/var/app/" + filename).readAllBytes());
    }

    public static void writeFile(String path, String content) throws IOException {
        new FileOutputStream("/uploads/" + path).write(content.getBytes());
    }

    // Weak cryptography (CWE-326, CWE-327)
    public static String hashPassword(String password) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("MD5");
        return Base64.getEncoder().encodeToString(md.digest(password.getBytes()));
    }

    public static String weakHash(String data) throws NoSuchAlgorithmException {
        MessageDigest md = MessageDigest.getInstance("SHA-1");
        return Base64.getEncoder().encodeToString(md.digest(data.getBytes()));
    }

    // Insecure random (CWE-330)
    public static String generateToken() {
        Random rand = new Random();
        return String.valueOf(rand.nextInt(999999));
    }

    public static int generateOTP() {
        return new Random().nextInt(9000) + 1000;
    }

    // XXE (CWE-611)
    public static void parseXML(String xml) throws Exception {
        javax.xml.parsers.DocumentBuilderFactory dbf =
            javax.xml.parsers.DocumentBuilderFactory.newInstance();
        // XXE not disabled
        javax.xml.parsers.DocumentBuilder db = dbf.newDocumentBuilder();
        db.parse(new ByteArrayInputStream(xml.getBytes()));
    }

    // Insecure deserialization (CWE-502)
    public static Object deserialize(byte[] data) throws Exception {
        ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(data));
        return ois.readObject();
    }

    // SSRF (CWE-918)
    public static String fetchURL(String url) throws IOException {
        return new URL(url).openConnection().getInputStream().readAllBytes().toString();
    }

    // Sensitive data exposure in logs (CWE-532)
    public static void login(String user, String password) {
        System.out.println("Login: user=" + user + " password=" + password);
    }

    // Null pointer dereference (CWE-476)
    public static int getLength(String s) {
        return s.length();  // No null check
    }

    // Stack overflow via recursion (CWE-674)
    public static int infiniteRecurse(int n) {
        return infiniteRecurse(n + 1);
    }

    public static void main(String[] args) throws Exception {
        getUser("admin' OR '1'='1'--");
        pingHost("google.com; cat /etc/passwd");
        readFile("../../etc/shadow");
        System.out.println(hashPassword("password"));
        System.out.println(generateToken());
    }
}

File: vulnerable_app.kt
import java.sql.DriverManager
import java.security.MessageDigest
import java.io.File
import java.io.ObjectInputStream
import java.io.ByteArrayInputStream
import java.net.URL
import java.util.Random

// Hardcoded credentials (CWE-798)
const val DB_PASSWORD = "kotlin_hardcoded_pass_123"
const val API_KEY = "kotlin_hardcoded_api_key_abcdef"
const val JWT_SECRET = "kotlin_jwt_secret_never_rotate"

// SQL Injection (CWE-89)
fun getUser(username: String) {
    val conn = DriverManager.getConnection("jdbc:mysql://localhost/db", "root", DB_PASSWORD)
    val stmt = conn.createStatement()
    stmt.executeQuery("SELECT * FROM users WHERE username = '$username'")
}

fun searchOrders(term: String) {
    val conn = DriverManager.getConnection("jdbc:mysql://localhost/db", "root", DB_PASSWORD)
    val stmt = conn.createStatement()
    stmt.executeQuery("SELECT * FROM orders WHERE item LIKE '%$term%'")
}

// Command Injection (CWE-78)
fun pingHost(host: String) {
    Runtime.getRuntime().exec(arrayOf("/bin/sh", "-c", "ping -c 1 $host"))
}

fun runScript(script: String) {
    Runtime.getRuntime().exec(script)
}

// Path Traversal (CWE-22)
fun readFile(filename: String): String {
    return File("/var/app/$filename").readText()
}

fun writeFile(path: String, content: String) {
    File("/uploads/$path").writeText(content)
}

// Weak cryptography (CWE-326)
fun hashPassword(password: String): String {
    val md = MessageDigest.getInstance("MD5")
    return md.digest(password.toByteArray()).joinToString("") { "%02x".format(it) }
}

fun weakHash(data: String): String {
    val sha1 = MessageDigest.getInstance("SHA-1")
    return sha1.digest(data.toByteArray()).joinToString("") { "%02x".format(it) }
}

// Insecure Deserialization (CWE-502)
fun deserialize(data: ByteArray): Any {
    val ois = ObjectInputStream(ByteArrayInputStream(data))
    return ois.readObject()
}

// SSRF (CWE-918)
fun fetchURL(url: String): String {
    return URL(url).readText()
}

// Insecure Random (CWE-330)
fun generateToken(): Int {
    return Random().nextInt(999999)
}

fun generateOTP(): String {
    return String.format("%04d", Random().nextInt(10000))
}

// XSS (CWE-79)
fun renderPage(userInput: String): String {
    return "<html><body>$userInput</body></html>"
}

// Sensitive data in logs (CWE-532)
fun login(user: String, password: String): Boolean {
    println("Login attempt: user=$user password=$password")
    return password == DB_PASSWORD
}

// Null pointer dereference (CWE-476)
fun getLength(s: String?): Int {
    return s!!.length   // Force unwrap - can throw NPE
}

fun main() {
    getUser("admin' OR '1'='1'--")
    pingHost("google.com; cat /etc/passwd")
    readFile("../../etc/shadow")
    println(hashPassword("password"))
    println(generateToken())
}
